#############################################################
#
# Ref to the ARTICLE
# 
#  Code to generate the Phyloseq objects and DESeq objects used in Maver et al., manuscript
#  Revision 04/21 d.bulgarelli@dundee.ac.uk  
#
#############################################################

#############################################################
# Clean-up the memory and start a new session
#############################################################

rm(list=ls())
dev.off()

#############################################################
# Libraries required
#############################################################

#required packages 
library("phyloseq")
library("DESeq2")
library ("ape")

#retrieve R and package versions and compare to the uploaded file in gtHub for the reproducibility of the code
sessionInfo()

#############################################################
#import the DADA2 phyloseq object generated by James Abbott
#############################################################

#JH12_ASV
JH12_phyloseq_DADA2 <- readRDS("JH12_dada2.rds")
JH12_phyloseq_DADA2 

##################################################################
#Pre-processing: remove Chloroplast and Mitochondria but retain NA
#################################################################

JH02_no_chlor <-subset_taxa(JH12_phyloseq_DADA2, (Order!="Chloroplast") | is.na(Order))
JH02_no_chlor

JH12_phyloseq_DADA2_no_plants <-subset_taxa(JH02_no_chlor, (Family!="Mitochondria") | is.na(Family))
JH12_phyloseq_DADA2_no_plants

##################################################################
#Prune putative contaminant ASVs
##################################################################

#Import the list of contaminat ASVs from JH06 library
Contaminant_ASVs <- read.delim("JH06_contaminant_ASVs_ids.txt", header = FALSE)

#identify the proportion of putative contaminants in the merged object
JH12_contaminants <- intersect(taxa_names(JH12_phyloseq_DADA2_no_plants),Contaminant_ASVs)
JH12_contaminants

#########################################################################
# Remove ASVs assigned to NA at phylum level
#########################################################################

JH12_phyloseq_DADA2_no_plants_1 <- subset_taxa(JH12_phyloseq_DADA2_no_plants, Phylum!= "NA")

#########################################################################
# Abundance threshold: 20 reads, set as 11% the minimum number of samples
########################################################################

JH12_phyloseq_DADA2_no_plants_2 = filter_taxa(JH12_phyloseq_DADA2_no_plants_1, function(x) sum(x > 20) > (0.02 *length(x)), TRUE)
JH12_phyloseq_DADA2_no_plants_2 
sort(sample_sums(JH12_phyloseq_DADA2_no_plants_2))

##ratio filtered reads/total reads
ratio <- sum(sample_sums(JH12_phyloseq_DADA2_no_plants_2))/sum(sample_sums(JH12_phyloseq_DADA2_no_plants))*100
ratio

#########################################################################
# Aggregate samples at genus level (Note: NArm set to false as Phylum NA ASVs pruned above)
########################################################################

JH12_data_phyloseq_genus <- tax_glom(JH12_phyloseq_DADA2_no_plants_2, taxrank= "Genus", NArm=FALSE, bad_empty=c(NA, "", " ", "\t"))

#compare the two objects
#ASVs
JH12_phyloseq_DADA2_no_plants_2
sort(sample_sums(JH12_phyloseq_DADA2_no_plants_2))
#Genera
JH12_data_phyloseq_genus
sort(sample_sums(JH12_data_phyloseq_genus))

####################################################################################
# Rarefy at an even sequencing depth (25,000) and "freeze" these objects for downstream analyses
#####################################################################################

#ASVs : ignore the warnings, the object will be saved right after
#JH12_rare_ASV_25K <- rarefy_even_depth(JH12_phyloseq_DADA2_no_plants_2, 25000)
#saveRDS(JH12_rare_ASV_25K, file ="JH12_data_phyloseq_ASV_0221_Silva138.rds")

#Genus : ignore the warnings, the object will be saved right after
#JH12_rare_genus_25K <- rarefy_even_depth(JH12_data_phyloseq_genus, 25000)
#saveRDS(JH12_rare_genus_25K, file ="JH12_data_phyloseq_genus_0221_Silva138.rds")

####################################################################################
# End

JH12_data_phyloseq_G0 <- subset_samples(JH12_data_phyloseq_genus, Treatments =="G0")
JH12_data_phyloseq_G24 <- subset_samples(JH12_data_phyloseq_genus, Treatments =="G24")
JH12_data_phyloseq_G46 <- subset_samples(JH12_data_phyloseq_genus, Treatments =="G46")

#G0
#extract count data 
JH12_counts_integer <- otu_table(JH12_data_phyloseq_G0)
countData = as.data.frame(JH12_counts_integer)
colnames(JH12_counts_integer)

#the design file containing sample information
colData = design[colnames(JH12_counts_integer), ]
rownames(colData)

#construct a DESeq dataset combining count data and sample information
JH12_G0_cds <- DESeqDataSetFromMatrix(countData =countData, colData=colData , design= ~Description)
saveRDS(JH12_G0_cds, file = "JH12_G0_genus_cds_0221.rds")

#G24
#extract count data 
JH12_counts_integer <- otu_table(JH12_data_phyloseq_G24)
countData = as.data.frame(JH12_counts_integer)
colnames(JH12_counts_integer)

#the design file containing sample information
colData = design[colnames(JH12_counts_integer), ]
rownames(colData)

#construct a DESeq dataset combining count data and sample information
JH12_G24_cds <- DESeqDataSetFromMatrix(countData =countData, colData=colData , design= ~Description)
saveRDS(JH12_G24_cds, file = "JH12_G24_genus_cds_0221.rds")

#G46
#extract count data 
JH12_counts_integer <- otu_table(JH12_data_phyloseq_G46)
countData = as.data.frame(JH12_counts_integer)
colnames(JH12_counts_integer)

#the design file containing sample information
colData = design[colnames(JH12_counts_integer), ]
rownames(colData)

#construct a DESeq dataset combining count data and sample information
JH12_G46_cds <- DESeqDataSetFromMatrix(countData =countData, colData=colData , design= ~Description)
saveRDS(JH12_G46_cds, file = "JH12_G46_genus_cds_0221.rds")

#end



#subset for Genotype
JH12_data_phyloseq_Morex <- subset_samples(JH12_data_phyloseq_genus, Description =="MOREXQUARRY")
JH12_data_phyloseq_Barke <- subset_samples(JH12_data_phyloseq_genus, Description  =="BARKEQUARRY")

#Morex
#extract count data 
JH12_counts_integer <- otu_table(JH12_data_phyloseq_Morex)
countData = as.data.frame(JH12_counts_integer)
colnames(JH12_counts_integer)

#the design file containing sample information
colData = design[colnames(JH12_counts_integer), ]
rownames(colData)

#construct a DESeq dataset combining count data and sample information
JH12_Morex_cds <- DESeqDataSetFromMatrix(countData =countData, colData=colData , design= ~Treatments)
saveRDS(JH12_Morex_cds, file = "JH12_Morex_cds_0221.rds")

#Barke
#extract count data 
JH12_counts_integer <- otu_table(JH12_data_phyloseq_Barke)
countData = as.data.frame(JH12_counts_integer)
colnames(JH12_counts_integer)

#the design file containing sample information
colData = design[colnames(JH12_counts_integer), ]
rownames(colData)

#construct a DESeq dataset combining count data and sample information
JH12_Barke_cds <- DESeqDataSetFromMatrix(countData =countData, colData=colData , design= ~Treatments)
saveRDS(JH12_Barke_cds, file = "JH12_Barke_cds_0221.rds")

